File: hashing/matrix/c++/set_zeros/set_zeros.cpp
// set_zeros.cpp
#include "set_zeros.hpp"
#include <vector>
#include <unordered_set>
#include <Eigen/Dense>

void SetZeros::hashApproach(std::vector<std::vector<int>> &matrix)
{
    int m = matrix.size(), n = matrix[0].size();
    std::unordered_set<int> rowsToZero;
    std::unordered_set<int> colsToZero;

    // First pass to identify rows and cols to zero
    for (int row = 0; row < m; row++)
    {
        for (int col = 0; col < n; col++)
        {
            if (matrix[row][col] == 0)
            {
                rowsToZero.insert(row);
                colsToZero.insert(col);
            }
        }
    }

    // Second pass to set elements to zeros
    for (int row = 0; row < m; row++)
    {
        for (int col = 0; col < n; col++)
        {
            if (rowsToZero.find(row) != rowsToZero.end() || colsToZero.find(col) != colsToZero.end())
            {
                matrix[row][col] = 0;
            }
        }
    }
}

void SetZeros::eigenApproach(Eigen::MatrixXd &matrix)
{
    int m = matrix.rows();
    int n = matrix.cols();
    bool firstRowZero = false;
    bool firstColZero = false;

    // Check if first row has zero O(n)
    for (int col = 0; col < n; col++)
    {
        if (matrix(0, col) == 0)
        {
            // Break as soon as we encounter a zero
            firstRowZero = true;
            break;
        }
    }

    // Check if first column has zero O(m)
    for (int row = 0; row < m; row++)
    {
        if (matrix(row, 0) == 0)
        {
            // Again, break as soon as we encounter a zero
            firstColZero = true;
            break;
        }
    }

    // Use first row and column to mark zeros
    for (int row = 1; row < m; row++)
    {
        for (int col = 1; col < n; col++)
        {
            // If a zero element is encountered
            if (matrix(row, col) == 0)
            {
                // Mark the entire column for zeroing
                matrix(row, 0) = 0;
                // Mark the entire row for zeroing
                matrix(0, col) = 0;
            }
        }
    }

    // Zero out cells based on marks in the first row and column
    for (int row = 1; row < m; row++)
    {
        if (matrix(row, 0) == 0)
        {
            matrix.row(row).setZero();
        }
    }

    for (int col = 1; col < n; ++col)
    {
        if (matrix(0, col) == 0)
        {
            matrix.col(col).setZero();
        }
    }

    // Zero out the first row if needed
    if (firstRowZero)
    {
        matrix.row(0).setZero();
    }

    // Zero out the first column if needed
    if (firstColZero)
    {
        matrix.col(0).setZero();
    }
}

File: hashing/matrix/python3/set_zeros.py
from typing import List, Set

import numpy as np


class SetZeros(object):
    """
    Given an `m x n` integer `matrix`, if an element is 0, set its entire row and column to 0's.

    Parameters
    ----------
    matrix : Union[List[List[int]], np.ndarray]
        A 2D matrix
    """

    @staticmethod
    def set_approach(matrix: List[List[int]]) -> None:
        rows_to_zero: Set[int] = set()
        cols_to_zero: Set[int] = set()
        m, n = len(matrix), len(matrix[0])
        # First pass to find all rows and columns that need to be zeroed 0(m x n)
        for row in range(m):
            for col in range(n):
                if matrix[row][col] == 0:
                    rows_to_zero.add(row)
                    cols_to_zero.add(col)
        # Second pass to set elements to zeros 0(m x n)
        for row in range(m):
            for col in range(n):
                if row in rows_to_zero or col in cols_to_zero:
                    matrix[row][col] = 0
        return None

    @staticmethod
    def numpy_approach(matrix: np.ndarray) -> None:
        # Find all zero elements, retruning (N, matrix.ndim) where N is the number of non-zero items
        zero_elements = np.argwhere(matrix == 0)
        # If no zero elements, do nothing
        if zero_elements.size == 0:
            return None
        # Determine the rows and columns to be zeroed
        rows_to_zero = np.unique(zero_elements[:, 0])
        cols_to_zero = np.unique(zero_elements[:, 1])
        # Zero out the rows and cols
        matrix[rows_to_zero, :] = 0
        matrix[:, cols_to_zero] = 0
        return None


def main() -> int:
    test_matrices = [
        [[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]],
        [[1, 1, 1], [1, 0, 1], [1, 1, 1]],
        [[0, 2, 7, 12], [4, 2, 0, 9], [5, 0, 2, 1]],
    ]

    for matrix_list in test_matrices:
        matrix = np.array(matrix_list)
        print(f"Matrix before setting elements to zero: \n{matrix}")
        SetZeros.numpy_approach(matrix)
        print(f"Matrix after setting elements to zero: \n {matrix}\n")

    return 0


if __name__ == "__main__":
    main()
